File: LeetCode/552. Student Attendance Record II/552.java
class Solution {
    public int checkRecord(int n) {
        int mod = 1000000007;
        int[][][] dp = new int[n + 1][2][3];
        dp[0][0][0] = 1;

        for (int i = 1; i <= n; i++) {
            for (int A = 0; A < 2; A++) {
                for (int L = 0; L < 3; L++) {
                    // Add 'P'
                    dp[i][A][0] = (dp[i][A][0] + dp[i - 1][A][L]) % mod;

                    // Add 'A'
                    if (A > 0) {
                        dp[i][A][0] = (dp[i][A][0] + dp[i - 1][A - 1][L]) % mod;
                    }

                    // Add 'L'
                    if (L > 0) {
                        dp[i][A][L] = (dp[i][A][L] + dp[i - 1][A][L - 1]) % mod;
                    }
                }
            }
        }

        int result = 0;
        for (int A = 0; A < 2; A++) {
            for (int L = 0; L < 3; L++) {
                result = (result + dp[n][A][L]) % mod;
            }
        }

        return result;
    }
}


/*
 class Solution {
    static final int M = 1000000007;

public int checkRecord(int n) {
    long[] PorL = new long[n + 1]; // ending with P or L, no A
    long[] P = new long[n + 1]; // ending with P, no A
    PorL[0] = P[0] = 1; PorL[1] = 2; P[1] = 1;

    for (int i = 2; i <= n; i++) {
        P[i] = PorL[i - 1];
        PorL[i] = (P[i] + P[i - 1] + P[i - 2]) % M;
    }
    
    long res = PorL[n];
    for (int i = 0; i < n; i++) { // inserting A into (n-1)-length strings
    	long s = (PorL[i] * PorL[n - i - 1]) % M;
        res = (res + s) % M;
    }
    
    return (int) res;
}
}
 */

File: LeetCode/552. Student Attendance Record II/552.py
class Solution(object):
    def checkRecord(self, n):
        """
        :type n: int
        :rtype: int
        """
        mod = 1000000007
        dp = [[[0] * 3 for _ in range(2)] for _ in range(n + 1)]
        dp[0][0][0] = 1

        for i in range(1, n + 1):
            for A in range(2):
                for L in range(3):
                    # Add 'P'
                    dp[i][A][0] = (dp[i][A][0] + dp[i - 1][A][L]) % mod

                    # Add 'A'
                    if A > 0:
                        dp[i][A][0] = (dp[i][A][0] + dp[i - 1][A - 1][L]) % mod

                    # Add 'L'
                    if L > 0:
                        dp[i][A][L] = (dp[i][A][L] + dp[i - 1][A][L - 1]) % mod

        result = 0
        for A in range(2):
            for L in range(3):
                result = (result + dp[n][A][L]) % mod

        return result
    



    '''
    class Solution(object):
    def checkRecord(self, n):
        """
        :type n: int
        :rtype: int
        """
        M = (10 ** 9) + 7
        l0a0 = 1
        l1a0 = 0
        l2a0 = 0
        l0a1 = 0
        l1a1 = 0
        l2a1 = 0

        for _ in range(n):
            l0a0, l0a1, l1a0, l2a0,  l1a1, l2a1 = (l0a0 + l1a0 + l2a0) % M,  (l0a1 + l1a1 + l2a1 + l0a0 + l1a0 + l2a0) % M, l0a0, l1a0, l0a1, l1a1
         
        return (l0a0 + l1a0 + l2a0 + l0a1 + l1a1 + l2a1) % M
    '''